### PR TITLE
Add pipeline for multi-speaker Parakeet RNNT

### PR DESCRIPTION
**Pull request by kingformatty**
Integrate speaker utilities into Parakeet RNNT Multi-speaker training/evaluation

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 1]
```diff
@@ -48,22 +52,84 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
+        
+        if 'diar_model_path' in self.cfg:
+            self.diar = True
+            # Initialize the speaker branch
+            self._init_diar_model()
+
+            if 'num_speaker' in cfg.model_defaults:
+                self.num_speakers = cfg.model_defaults.num_speakers
+            else:
+                self.num_speakers = 4
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]: This can be replaced by `self.num_speakers = cfg.get('num_speaker', 4)`
Actually it was me that coded in this way long time ago... and let's change it in this PR.

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 2]
```diff
@@ -48,22 +52,84 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
+            
+            # layer normalization, ln, l2, or None
+            if 'norm' in cfg:
+                if cfg.norm == 'ln':
+                    self.asr_norm = torch.nn.LayerNorm(cfg.model_defaults.enc_hidden)
+                    self.diar_norm = torch.nn.LayerNorm(4)
+                self.norm = cfg.norm
+            else:
+                self.norm = None
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]:
```
self.norm = cfg.get('norm', None)
if self.norm == 'ln':
    self.asr_norm = torch.nn.LayerNorm(cfg.model_defaults.enc_hidden)
    self.diar_norm = torch.nn.LayerNorm(4)
```

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 3]
```diff
@@ -48,22 +52,84 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
+
+            if 'kernel_norm' in cfg:
+                self.kernel_norm = cfg.kernel_norm
+            else:
+                self.kernel_norm = None
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]: `self.kernel_norm = cfg.get('kernel_norm', None)`

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 4]
```diff
@@ -48,22 +52,84 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
+
+            # projection layer
+            proj_in_size = 4 + cfg.model_defaults.enc_hidden
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]: Should be `proj_in_size = self.num_speakers + cfg.model_defaults.enc_hidden`

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 5]
```diff
@@ -48,22 +52,84 @@ def list_available_models(cls) -> List[PretrainedModelInfo]:
...
+                    self.diar_kernel_type = cfg.diar_kernel_type
+                    # projection layer
+                    proj_in_size = 4 * cfg.model_defaults.enc_hidden
+                    proj_out_size = cfg.model_defaults.enc_hidden
+                    self.joint_proj = torch.nn.Sequential(
+                        torch.nn.Linear(proj_in_size, proj_out_size*2),
+                        torch.nn.ReLU(),
+                        torch.nn.Linear(proj_out_size*2, proj_out_size)
+                    )
+                    self.diar_kernel = self.joint_proj
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]:
```
self.diar_kernel_type = cfg.get('diar_kernel_type', None)
if self.diar_kernel_type == 'sinusoidal':
    self.diar_kernel = self.get_sinusoid_position_encoding(self.num_speakers, cfg.model_defaults.enc_hidden)
elif self.diar_kernel_type == 'metacat':
    # projection layer
    proj_in_size = self.num_speakers * cfg.model_defaults.enc_hidden
    proj_out_size = cfg.model_defaults.enc_hidden
    self.joint_proj = torch.nn.Sequential(
        torch.nn.Linear(proj_in_size, proj_out_size*2),
        torch.nn.ReLU(),
        torch.nn.Linear(proj_out_size*2, proj_out_size)
    )
    self.diar_kernel = self.joint_proj
```

Please check if I made mistakes or typos.

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 6]
```diff
@@ -115,14 +177,89 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
...
+    def forward_diar(
+        self,
+        input_signal=None,
+        input_signal_length=None,
+    ):
+        preds, _preds, attn_score_stack, total_memory_list, encoder_states_list = self.diarization_model.forward(audio_signal=input_signal, audio_signal_length=input_signal_length)
+
+        return preds, _preds, attn_score_stack, total_memory_list, encoder_states_list
+
+    def fix_diar_output(
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py:199]: Do we still need this function as you implement that get_hidden_length function to get the actual target length?

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 7]
```diff
@@ -138,11 +275,64 @@ def training_step(self, batch, batch_nb):
...
+                    diar_preds = spk_targets 
+                else:
+                    raise ValueError("`spk_targets` is required for speaker supervision strategy 'rttm'")
+            elif self.cfg.spk_supervision_strategy == 'diar':
+                if diar_preds is None:
+                    raise ValueError("`diar_pred`is required for speaker supervision strategy 'diar'")
+            elif self.cfg.spk_supervision_strategy == 'mix':
+                diar_preds = self._get_probablistic_mix(diar_preds=diar_preds, spk_targets=spk_targets, rttm_mix_prob=float(self.cfg.rttm_mix_prob))
+            else:
+                raise ValueError(f"Invalid RTTM strategy {self.cfg.spk_supervision_strategy} is not supported.")
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]: Can we only calculate `diar_preds` for 'diar' and 'mix'? Because for the case where `spk_supervision_strategy=rttm`, we still calculate the diarization result.

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 8]
```diff
@@ -138,11 +275,64 @@ def training_step(self, batch, batch_nb):
...
+        
+            # Speaker mapping shuffling to equalize the speaker label's distributions
+            if self.cfg.shuffle_spk_mapping:
+                diar_preds = apply_spk_mapping(diar_preds, spk_mappings)
+
+            if(diar_preds.shape[1]!=encoded.shape[1]):
+            # KD duct-tape solution for extending the speaker predictions 
+                asr_frame_count = encoded.shape[1]
+                diar_preds = self.fix_diar_output(diar_preds, asr_frame_count)
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py:302]: Again, same as previous question, do we still have this frame mismatch issue?

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 9]
```diff
@@ -232,17 +422,69 @@ def training_step(self, batch, batch_nb):
     def validation_pass(self, batch, batch_idx, dataloader_idx=0):
```
Comment (weiqingw4ng) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py:428]: In my understanding, both `training_step` and `validation_pass` only take the results from `forward` and `decoder` function and do loss/wer calculation. All of the model calculation should be done in `forward` function so that you don't need to have duplicate code in `training_step` and `validation_pass`. I am not sure if it is already there when you got the code from our branch. But if you take a look at the `EncDecMultiTaskModel` in `nemo/collections/asr/models/aed_multitask_models.py`, actually you only need to change `forward` and keep the `training_step` and `validation_pass` unchanged. 

That would also make this code a lot simple, otherwise you have to change the code twice every time you have some new ideas. 

In my suggestion, you can have a new `forward_asr` function, which is exactly doing the same thing as the `forward` function does in `rnnt_model`. And you already have a `forward_diar` function. Then you can put everything like concatenation and projection in the `training_step` to the `forward` function and return `encoded`. In this way you don't need to change the `training_step` and `validation_pass` function in the original `rnnt_model` class.

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py (hunk 5 above)
Comment (kingformatty) [reply nemo/collections/asr/models/rnnt_spk_bpe_models.py]: Leave the projection-module definition for now. These are necessary to load previous checkpoint. Will remove it in future.

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py (hunk 6 above)
Comment (kingformatty) [reply nemo/collections/asr/models/rnnt_spk_bpe_models.py:199]: I think this is meant to resolve difference between asr encoder vs diar_preds difference. diar_preds are not from get_hidden_length_function, but from sortformer.
